plugins/diff/erofs/differ_linux.go

```diff
@@ -54,10 +61,151 @@ func NewErofsDiffer(store content.Store, mkfsExtraOpts []string) differ {
 	}
 }
 
+func writeDiff(ctx context.Context, w io.Writer, lower []mount.Mount, upperRoot string) error {
+	var opts []archive.ChangeWriterOpt
+
+	return mount.WithTempMount(ctx, lower, func(lowerRoot string) error {
+		cw := archive.NewChangeWriter(w, upperRoot, opts...)
+		err := fs.DiffDirChanges(ctx, lowerRoot, upperRoot, fs.DiffSourceOverlayFS, cw.HandleChange)
+		if err != nil {
+			return fmt.Errorf("failed to create diff tar stream: %w", err)
+		}
+		return cw.Close()
+	})
+}
+
 // Compare creates a diff between the given mounts and uploads the result
 // to the content store.
 func (s erofsDiff) Compare(ctx context.Context, lower, upper []mount.Mount, opts ...diff.Opt) (d ocispec.Descriptor, err error) {
```

Comment:
What's the main difference between this and `walking.Compare` (https://github.com/containerd/containerd/blob/b22a302a75d9a7d7955780e54cc5b32de6c8525d/plugins/diff/walking/differ.go#L62)?

Is it only changing `archive.WriteDiff` to `writeDiff`?

Reply:
yes, archive.WriteDiff applies to all snapshotters, but since erofs differ ties to erofs snapshotter (erofs uses overlayfs for writable layer), so it's no needed to compare the layers, but using overlayfs upperdir directly.